lighthouse-core/gather/computed/main-thread-tasks.js

```diff
@@ -0,0 +1,215 @@
+        currentTask.endTime <= event.ts
+      ) {
+        currentTask = currentTask.parent;
+      }
+
+      // If we don't have a current task, start a new one.
+      if (!currentTask) {
+        // We can't start a task with an end event
+        if (event.ph === 'E') {
+          throw new Error('Fatal trace logic error');
+        }
+
+        currentTask = MainThreadTasks._createNewTaskNode(event);
+        tasks.push(currentTask);
+
+        continue;
+      }
+
+      if (event.ph === 'X' || event.ph === 'B') {
+        // We're starting a nested event, create it as a child and make it the currentTask
+        const newTask = MainThreadTasks._createNewTaskNode(event, currentTask);
+        tasks.push(newTask);
+        currentTask = newTask;
+      } else {
+        if (currentTask.event.ph !== 'B') {
+          throw new Error('Fatal trace logic error');
+        }
+
+        // We're ending an event, update the end time and the currentTask to its parent
+        currentTask.endTime = event.ts;
+        currentTask = currentTask.parent;
+      }
+    }
+
+    return tasks;
+  }
+
+  /**
+   * @param {TaskNode} task
+   * @return {number}
+   */
+  static _computeRecursiveSelfTime(task) {
+    const childTime = task.children
+      .map(MainThreadTasks._computeRecursiveSelfTime)
+      .reduce((sum, child) => sum + child, 0);
+    task.duration = task.endTime - task.startTime;
+    task.selfTime = task.duration - childTime;
+    return task.duration;
+  }
+
+  /**
+   * @param {TaskNode} task
+   * @param {string} [parentURL]
+   */
+  static _computeRecursiveAttributableURL(task, parentURL) {
+    const argsData = task.event.args.data || {};
+    const stackFrames = argsData.stackTrace || [{url: undefined}];
+    const taskURL = argsData.url || (stackFrames[0] && stackFrames[0].url);
+
+    task.attributableURL = parentURL || taskURL;
+    task.children.forEach(child =>
+      MainThreadTasks._computeRecursiveAttributableURL(child, task.attributableURL));
+  }
+
+  /**
+   * @param {TaskNode} task
+   * @param {TaskGroup} [parentGroup]
+   */
+  static _computeRecursiveTaskGroup(task, parentGroup) {
+    const group = taskNameToGroup[task.event.name];
+    task.group = group || parentGroup || taskGroups.other;
+    task.children.forEach(child => MainThreadTasks._computeRecursiveTaskGroup(child, task.group));
+  }
+
+  /**
+   *
+   * @param {LH.TraceEvent[]} traceEvents
+   * @return {TaskNode[]}
+   */
+  static getMainThreadTasks(traceEvents) {
+    const tasks = MainThreadTasks._createTasksFromEvents(traceEvents);
+
+    // Compute the recursive properties we couldn't compute earlier, starting at the toplevel tasks
+    for (const task of tasks) {
+      if (task.parent) continue;
+
+      MainThreadTasks._computeRecursiveSelfTime(task);
+      MainThreadTasks._computeRecursiveAttributableURL(task);
+      MainThreadTasks._computeRecursiveTaskGroup(task);
+    }
+
+    const firstTs = (tasks[0] || {startTime: 0}).startTime;
```

Comment:
i've always heard the traceEvents could come out of order, however I know this is more of a case of cross-process events. Perhaps not an issue if everything is within the same thread?  We should doublecheck with caseq.

Reply:
oh yeah, we should make sure to use the sorted events here 👍

Reply:
> oh yeah, we should make sure to use the sorted events here 👍

should this depend on `traceOfTab.mainThreadEvents`? Then it doesn't need to check if tasks are in the main thread and they're already sorted

Reply:
yup that was my plan 👍